Include/Rover/LinearRegressionModel.hpp

```diff
@@ -0,0 +1,73 @@
+#ifndef ROVER_LINEAR_REGRESSION_MODEL_HPP
+#define ROVER_LINEAR_REGRESSION_MODEL_HPP
+#include <dlib/matrix.h>
+
+namespace Rover {
+  //! Models a trial using linear regression.
+  /*!
+    \tparam T The Trial to model.
+  */
+  template<typename T>
+  class LinearRegressionModel {
+    public:
+
+      //! The Trial to model.
+      using Trial = T;
+
+      //! The result of the trial.
+      using Result = typename Trial::Sample::Result;
+
+      //! Constructs a linear regression model for a trial.
+      /*!
+        \param trial The trial to model.
+      */
+      LinearRegressionModel(const Trial& trial);
+
+      template<typename... Arg>
+      Result operator ()(Arg&&... arg) const;
+
+    private:
+      static dlib::matrix<Result> compute_transformation_vector(const T& trial);
+
+    private:
+      dlib::matrix<Result> m_transformation;
+  };
+
+  template<typename T>
+  LinearRegressionModel<T>::LinearRegressionModel(const T& trial)
+    : m_transformation(compute_transformation_vector(trial)) {}
+
+  template<typename T>
+  template<typename... Arg>
+  typename LinearRegressionModel<T>::Result
+      LinearRegressionModel<T>::operator ()(Arg&&... arg) const {
+    auto x = dlib::matrix<Result, 1, std::tuple_size_v<
+      typename Trial::Sample::Parameters> + 1>();
+    x = { 1., std::forward<Arg>(arg)... };
+    auto result = static_cast<Result>(x * m_transformation);
+    return result;
+  }
+
+  template<typename T>
+  dlib::matrix<typename LinearRegressionModel<T>::Result> 
+      LinearRegressionModel<T>::compute_transformation_vector(const T& trial) {
+    auto x = dlib::matrix<Result>(trial.size(),
+      std::tuple_size_v<typename T::Sample::Parameters> + 1);
+    for(auto i = 0U; i < trial.size(); ++i) {
+      auto row = std::apply([](const auto&... param) {
+        return std::vector<Result> { 1., param... };
+      }, trial[i].m_arguments);
+      std::move(row.begin(), row.end(), x.begin() + i * x.nc());
```

Comment:
Would ```x.end()``` work?

Reply:
matrix<T>::end() returns an element-wise pte iterator (i.e the element after the last on the last row), and begin() is an iterator to the first element of the first row. x.begin() + i * x.nc(), where nc() is the number of columns, is an iterator to the first element of the ith row. There is no way to iterate directly over rows, which is somewhat inconvenient. So I basically have to stride myself.